File: find_the_original_typed_string_II.js
class Solution {
  possibleStringCount(word, k) {
    const MOD = 1e9 + 7;

    if (!word) return 0;

    let groups = [];
    let consecutiveCharactersCount = 1;

    for (let i = 1; i < word.length; i++) {
      if (word[i] === word[i - 1]) {
        consecutiveCharactersCount++;
      } else {
        groups.push(consecutiveCharactersCount);
        consecutiveCharactersCount = 1;
      }
    }
    groups.push(consecutiveCharactersCount);

    let totalPossibleCombinations = 1;
    for (const frequency of groups) {
      totalPossibleCombinations = (totalPossibleCombinations * frequency) % MOD;
    }

    if (k <= groups.length) return totalPossibleCombinations;

    let prev = Array(k).fill(0);
    prev[0] = 1;

    for (const count of groups) {
      let curr = Array(k).fill(0);
      let currSum = 0;
      for (let i = 0; i < k; i++) {
        if (i > 0) currSum = (currSum + prev[i - 1]) % MOD;
        if (i > count) currSum = (currSum - prev[i - count - 1] + MOD) % MOD;
        curr[i] = currSum;
      }
      prev = curr;
    }

    const invalidCombinations = prev.slice(groups.length).reduce((a, b) => (a + b) % MOD, 0);
    return (totalPossibleCombinations - invalidCombinations + MOD) % MOD;
  }
}

File: find_the_original_typed_string_II.py
# Time Complexity: O(N+K^2)  |  Space Complexity: O(K)

class Solution:

    def possibleStringCount(self, word: str, k: int) -> int:
        MOD = 10**9+7
        if not word:
            return 0

        groups = []
        consecutiveCharactersCount = 1
        for i in range(1, len(word)):
            if word[i] == word[i - 1]:
                consecutiveCharactersCount += 1
            else:
                groups.append(consecutiveCharactersCount)
                consecutiveCharactersCount = 1
        groups.append(consecutiveCharactersCount)

        totalPossibleCombinations = 1
        for frequency in groups:
            totalPossibleCombinations = (totalPossibleCombinations * frequency) % MOD

        if k <= len(groups):
            return totalPossibleCombinations

        prev = [0] * k
        prev[0] = 1

        for count in groups:
            curr = [0] * k
            currSum = 0
            for i in range(k):
                if i > 0:
                    currSum = (currSum+ prev[i-1]) % MOD
                if i > count:
                    currSum = (currSum - prev[i - count - 1] + MOD) % MOD
                curr[i] = currSum
            prev = curr

        invalidCombinations = sum(curr[len(groups):k]) % MOD
        return (totalPossibleCombinations - invalidCombinations + MOD) % MOD
